src/Dictionaries/LibraryDictionarySource.cpp

```diff
@@ -8,292 +9,182 @@
+{
+    bridge_helper->removeLibrary();
+}
+
+
 LibraryDictionarySource::LibraryDictionarySource(const LibraryDictionarySource & other)
     : log(&Poco::Logger::get("LibraryDictionarySource"))
     , dict_struct{other.dict_struct}
     , config_prefix{other.config_prefix}
     , path{other.path}
+    , dictionary_id{createDictID()}
     , sample_block{other.sample_block}
-    , library{other.library}
+    , context(other.context)
     , description{other.description}
-    , settings{other.settings}
 {
-    if (auto lib_clone = library->tryGet<ClickHouseLibrary::LibraryCloneFunc>(ClickHouseLibrary::LIBRARY_CLONE_FUNC_NAME))
-        lib_data = lib_clone(other.lib_data);
-    else if (auto lib_new = library->tryGet<ClickHouseLibrary::LibraryNewFunc>(ClickHouseLibrary::LIBRARY_CREATE_NEW_FUNC_NAME))
-        lib_data = lib_new(&settings->strings, ClickHouseLibrary::log);
+    bridge_helper = std::make_shared<LibraryBridgeHelper>(context, dictionary_id);
+    bridge_helper->cloneLibrary(other.dictionary_id);
 }
 
-LibraryDictionarySource::~LibraryDictionarySource()
-{
-    if (auto lib_delete = library->tryGet<ClickHouseLibrary::LibraryDeleteFunc>(ClickHouseLibrary::LIBRARY_DELETE_FUNC_NAME))
-        lib_delete(lib_data);
-}
 
-BlockInputStreamPtr LibraryDictionarySource::loadAll()
+bool LibraryDictionarySource::isModified() const
 {
-    LOG_TRACE(log, "loadAll {}", toString());
-
-    auto columns_holder = std::make_unique<ClickHouseLibrary::CString[]>(dict_struct.attributes.size());
-    ClickHouseLibrary::CStrings columns{static_cast<decltype(ClickHouseLibrary::CStrings::data)>(columns_holder.get()),
-                                        dict_struct.attributes.size()};
-    size_t i = 0;
-    for (const auto & a : dict_struct.attributes)
-    {
-        columns.data[i] = a.name.c_str();
-        ++i;
-    }
+    return bridge_helper->isModified();
+}
 
-    auto load_all_func = library->get<ClickHouseLibrary::LibraryLoadAllFunc>(ClickHouseLibrary::LIBRARY_LOAD_ALL_FUNC_NAME);
-    auto data_new_func = library->get<ClickHouseLibrary::LibraryDataNewFunc>(ClickHouseLibrary::LIBRARY_DATA_NEW_FUNC_NAME);
-    auto data_delete_func = library->get<ClickHouseLibrary::LibraryDataDeleteFunc>(ClickHouseLibrary::LIBRARY_DATA_DELETE_FUNC_NAME);
 
-    ClickHouseLibrary::LibraryData data_ptr = data_new_func(lib_data);
-    SCOPE_EXIT(data_delete_func(lib_data, data_ptr));
+bool LibraryDictionarySource::supportsSelectiveLoad() const
+{
+    return bridge_helper->supportsSelectiveLoad();
+}
 
-    ClickHouseLibrary::RawClickHouseLibraryTable data = load_all_func(data_ptr, &settings->strings, &columns);
-    auto block = dataToBlock(description.sample_block, data);
 
-    return std::make_shared<OneBlockInputStream>(block);
+BlockInputStreamPtr LibraryDictionarySource::loadAll()
+{
+    LOG_TRACE(log, "loadAll {}", toString());
+    return bridge_helper->loadAll(getDictAttributesString(), description.sample_block);
```

Comment:
You may notice that we don't need attributes because we have a sample block: https://github.com/ClickHouse/ClickHouse/blob/master/src/Dictionaries/DictionarySourceFactory.cpp#L53-L59